### ISSUE TITLE
Support update of "put" values

### ISSUE THREAD

**Issue by rhc54** [pmix/pmix#119]
Consider the use-case of two ranks that:
- each "put" a value with key "a", then put a value with key "b"
- call commit and fence
- retrieve the values for "a" and "b"
- update and "put" the value for key "a"
- call commit and fence
- retrieve the values for "a" and "b"

At the end, they will have the old values for "a", even though they were put/commit/fence'd, because the client will locally satisfy the "get" request. We need a mechanism by which the server can "invalidate" the client's local data store so it knows that it needs to re-pull the data from the server.

There are three options:
- automatically invalidate all locally cached data from procs involved in a fence. This would only alert procs involved in the immediate fence.
- pass down an invalidate cmd in the fence return to indicate which procs have changed their info. This would also only alert procs involved in the immediate fence.
- use the notification system to alert procs that someone modified their "put" data. This would alert even those procs that were not involved in a subsequent fence so they could update their info - e.g., when a proc is restarted.

@dsolt @artpol84 @bosilca @abouteiller Any thoughts/preferences?

**Comment by rhc54** [comment]
Adding @afriedle-intel

**Comment by abouteiller** [comment]
Option 1: Invalidating the entire cache on every fence seems excessive. 
Option 3: notification is not synchronous with respect to fence, is it? I'd expect this to suffer from race conditions.

**Comment by bosilca** [comment]
I would advocate for partial invalidation supported by data versioning ? Each put modify the local version of the data (monotonically increasing it's version so that we can drop late messages with older versions). When the data is transmitted, it's version is also provided, so that the peers know not only the content of the data but also it's version.

On subsequent fences each process will mark all known peers data as potentially invalid (but will not modify the version number yet). On a get where the proc has been marked as dirty, the server will recheck if the peer data has been updated. As we version the data, this check can be as simple as "is proc X data still at version Y?"

Using the versioning we can also imagine that the fence become a allgather of all versions, so that we can mark as dirty only the procs with updated keys.

**Comment by rhc54** [comment]
Do we need to retain the version info? I ask only from the standpoint of memory footprint. It isn't clear to me what the version does for us - are we actually anticipating a use for the version number?

We know if a "commit" actually added or changed any existing info, and so a fence that collects data can also know if something was updated (we can just flag that a contribution from a given proc is an updated blob). So fence recipients will know if data from a particular proc was updated or not.

Would this suffice? Or do they really need to know how many times it has been changed?

**Comment by bosilca** [comment]
The version can be as short as 16 bits, we just need it not to overflow during the run. The version will allow us to minimize the amount of data communicated, or to even skip the communication completely (if the fence is doing the allgather).

My understanding is that we need the version to be able to cope with the fact that the data on different peers is out of sync. At least this will provide more flexibility on what we do during the fence.

However, you might be right. A specific implementation of the fence can handle everything with a simple state flag (updated or not). This will also require all servers to keep the update flag for all peers, and to update the local knowledge accordingly (if the remote data has been updated, then mark the local cache as dirty; if the remote data has not been updated but the local data is marked as dirty it should remain as such).

**Comment by artpol84** [comment]
1. I was rising key versioning question some time ago and the answer was - "this is not a typical PMIx usage, we don't intend to optimize/fully support this (in case of the direct modex". **What has changed now**?
2. FWIW This will not be a problem with the dstor. Dstor has the option to invalidate the key.

@bosilca 
"... *On a get where the proc has been marked as dirty, the server will recheck if the peer data has been updated. As we version the data, this check can be as simple as "is proc X data still at version Y?"* ..."

As far as I know in case you don't use dstor PMIx server doesn't deal with keys, but with the rank BLOBs. this means that server has no way of saying if the particular key has changed. If I understand correctly this will turn in to an "option 1 - invalidating the cache"

**Comment by rhc54** [comment]
Just to be clear: this still isn't a typical usage. The issue arose thru a deliberately crafted test where someone ran two multi-threaded procs, each banging put/commit/fence calls that re-used the keys. When the values didn't change, the test erroneously concluded that PMIx had a threading issue.

Changing the test so the keys themselves were updated as well as the values resolved the test. However, it raised the question about memory footprint - in some use-case (still to be imagined) where updating a key-value pair was required, then the need to continually update the key would cause the footprint to grow and raises questions about how the app knows which key to retrieve for a given purpose.

So the use-case motivating this question does not require versioning - it does, however, require that the client "know" that it needs to refresh either the job-level info or the blob for that proc (whichever has changed). Thus, a simple invalidation protocol would suffice to resolve this specific issue.

I think @bosilca is raising the question of whether or not the addition of versioning would buy us something in addition to solving the immediate issue. I'm not sure what that is just yet, but am open to hearing thoughts on the matter.

**Comment by rhc54** [comment]
Per telecon on Aug 4th:

  * We discussed this at some length. The primary concern was over just how much of a use-case we really have for this feature, especially recalling that the objective of PMIx is to eliminate the exchange of such information. Rather than make signficant changes to the code base, we propose instead that we add an info key PMIX\_REFRESH\_CACHE that can be passed as a directive to the PMIx\_Get call to indicate that the local cache is to be refreshed.

  * Our rationale is that the application must have some reason to believe that "put" information may have been updated. This is a rare use-case, if it even exists, and therefore can be treated as an exception.
